miniwin/miniwin/src/include/miniwin_d3drmmesh_p.h

```diff
@@ -2,33 +2,67 @@
 
 #include "miniwin_d3drmobject_p.h"
 
+#include <algorithm>
+#include <vector>
+
+struct MeshGroup {
+	D3DCOLOR color = 0xFFFFFFFF;
+	IDirect3DRMTexture* texture = nullptr;
+	IDirect3DRMMaterial* material = nullptr;
+	D3DRMRENDERQUALITY quality = D3DRMRENDER_GOURAUD;
+	int vertexPerFace = 0;
+	std::vector<D3DRMVERTEX> vertices;
+	std::vector<unsigned int> faces;
+
+	MeshGroup() = default;
+
+	MeshGroup(const MeshGroup& other)
+		: color(other.color), texture(other.texture), material(other.material), quality(other.quality),
+		  vertexPerFace(other.vertexPerFace), vertices(other.vertices), faces(other.faces)
+	{
+		if (texture) {
+			texture->AddRef();
+		}
+		if (material) {
+			material->AddRef();
+		}
+	}
+
+	// Move constructor
+	MeshGroup(MeshGroup&& other) noexcept = default;
+
+	// Move assignment
+	MeshGroup& operator=(MeshGroup&& other) noexcept = default;
+
+	~MeshGroup() {}
```

Comment:
The move constructor should copy all attributes, and leave the copied-from object in a valid state.
By settings its texture and material to null, we make sure its destructor will not Release them.
So their refcount remains identical.

```suggestion
	// Move constructor
	MeshGroup(MeshGroup&& other) noexcept 	
		: color(other.color), texture(other.texture), material(other.material), quality(other.quality),
		  vertexPerFace(other.vertexPerFace), vertices(other.vertices), faces(other.faces)
	{
		other->texture = nullptr;
		other->material = nullptr;
	}

	// Move assignment
	MeshGroup& operator=(MeshGroup&& other) noexcept {
		color = other.color;
		texture = other.texture;
		material = other.material;
		quality = other.quality;
		vertexPerFace = other.vertexPerFace;
		vertices = other.vertices;
		faces = other.faces;

		other.texture = nullptr;
		other.material = nullptr;
	}

	~MeshGroup() {
		if (texture) {
			texture->Release();
		}
		if (material) {
			material->Release();
		}
	}
```

Reply:
Wouldn't you want to move its resources like vertices and faces as well?

Reply:
Thanks I could not make it work, I think i was doing Release in the assignment incorrectly.